File: LEETCODE/Graphs/traversalProblems/S1_NumberOfProvience.java
package LEETCODE.Graphs.traversalProblems;

import java.util.*;

public class S1_NumberOfProvience {


    private static void bfs(int node, int[][] connected, boolean[] vis, int col){
        Queue<Integer> q = new LinkedList<>();
        q.offer(node);
        vis[node] = true;
        while(!q.isEmpty()){
            int nde = q.poll();

            for(int i  = 0; i < col; i++){
                // ab agar mera node visited nhi hai && or agar mera next neighbour connected neighbout 
                // curr = 1 ab mera 1 kis kis se conncted hai to merko har col mei check krna hai 
                if(!vis[i] && connected[nde][i] == 1){ // agar meri node visited nhi hai to 
                    vis[i] = true;
                    q.offer(i);
                }
            }
        }
    }
    public static  int findCircleNum(int[][] isConnected) {
        int v = isConnected.length; // total number of nodes 
        int col = isConnected[0].length;
        int  p = 0;
        boolean vis[] = new boolean[v];

        for(int i = 0; i < v; i++){
            if(vis[i] == false){
                p++;
                bfs(i, isConnected, vis, col);
            }
        }
        return p;
    }

    public static void main(String[] args) {
        int[][] isConnected = {
                                {1, 1, 0, 0, 0},
                                {1, 1, 1, 0, 0},
                                {0, 1, 1, 0, 0},
                                {0, 0, 0, 1, 1},
                                {0, 0, 0, 1, 1}
                            };
                            // 0 ---- 1 ---- 2

                            // 3 ---- 4
                            findCircleNum(isConnected);

    }
}

File: LEETCODE/Graphs/traversalProblems/S1_NumberOfProvience.cpp
#include <bits/stdc++.h>
using namespace std;

void dfs(int node, vector<vector<int>>& matrix, vector<int>& vis) {
    vis[node] = 1;
    for(int i = 0; i < matrix[node].size(); i++) {
        if(matrix[node][i] == 1 && vis[i] == 0) {
            dfs(i, matrix, vis);
        }
    }
}
int findCircleNum(vector<vector<int>>& isConnected) {
    int v = isConnected.size();
    vector<int> vis(v, 0);
    int count = 0;
    for(int i = 0; i < v; i++) {
        if(vis[i] == 0) {
            count++;
            dfs(i, isConnected, vis);
        }
    }

    return count;
}
int  main() {
    vector<vector<int>>isConnected = 
    {{1, 1, 0, 0, 0},
    {1, 1, 1, 0, 0},
    {0, 1, 1, 0, 0},
    {0, 0, 0, 1, 1},
    {0, 0, 0, 1, 1}};
    return 0;           
}

